File: backend-flask-pytorch/video/sources/start_time.py
"""Parse clip start times from form values and YouTube links."""

import math
import re
from urllib.parse import parse_qs, urlsplit

INVALID_TIME = "Use seconds, mm:ss, or hh:mm:ss for the start time."


def parse_seconds(text: str) -> float:
    """Read a non-negative number of seconds."""
    if not re.fullmatch(r"\d+(?:\.\d+)?", text):
        raise ValueError(INVALID_TIME)
    return float(text)


def parse_clock_time(text: str) -> float:
    """Read mm:ss or hh:mm:ss with valid minute and second fields."""
    if not re.fullmatch(r"\d+(?::\d{1,2}){1,2}(?:\.\d+)?", text):
        raise ValueError(INVALID_TIME)
    parts = [float(part) for part in text.split(":")]
    if any(part >= 60 for part in parts[1:]):
        raise ValueError(INVALID_TIME)
    seconds = 0.0
    for part in parts:
        seconds = seconds * 60 + part
    return seconds


def parse_youtube_time(text: str) -> float:
    """Read YouTube's optional h, m, and s suffixes."""
    match = re.fullmatch(r"(?:(\d+)h)?(?:(\d+)m)?(?:(\d+(?:\.\d+)?)s)?", text)
    if not match or not any(match.groups()):
        raise ValueError(INVALID_TIME)
    hours, minutes, seconds = (float(part or 0) for part in match.groups())
    return hours * 3600 + minutes * 60 + seconds


def validate_seconds(seconds: float) -> float:
    """Reject non-finite input and keep millisecond precision."""
    if not math.isfinite(seconds):
        raise ValueError("Enter a finite, non-negative start time.")
    return round(seconds, 3)


def parse_start_time(value: str | float | None) -> float:
    """Convert a supported timestamp to seconds; empty input starts at zero."""
    text = str(value).strip() if value is not None else ""
    if not text:
        return 0.0
    if ":" in text:
        return validate_seconds(parse_clock_time(text))
    if text.endswith(("h", "m", "s")):
        return validate_seconds(parse_youtube_time(text))
    return validate_seconds(parse_seconds(text))


def clip_start(data: dict, source: str, url: str) -> float:
    """Prefer an explicit form value over the YouTube link timestamp."""
    if data.get("startSeconds") not in (None, ""):
        return parse_start_time(data["startSeconds"])
    if source != "youtube":
        return 0.0
    parsed = urlsplit(url)
    query = parse_qs(parsed.query)
    fragment = parse_qs(parsed.fragment)
    timestamp = query.get("t") or query.get("start") or fragment.get("t")
    return parse_start_time(timestamp[0]) if timestamp else 0.0

File: frontend/src/assets/components/video/helpers/startTime.ts
const INVALID_TIME = "Use seconds, mm:ss, or hh:mm:ss for the start time.";

const parseSeconds = (text: string) => {
  if (!/^\d+(?:\.\d+)?$/.test(text)) throw new Error(INVALID_TIME);
  return Number(text);
};

const parseClockTime = (text: string) => {
  if (!/^\d+(?::\d{1,2}){1,2}(?:\.\d+)?$/.test(text)) {
    throw new Error(INVALID_TIME);
  }
  const parts = text.split(":").map(Number);
  if (parts.slice(1).some((part) => part >= 60)) throw new Error(INVALID_TIME);
  return parts.reduce((total, part) => total * 60 + part, 0);
};

const parseYoutubeTime = (text: string) => {
  const match = /^(?:(\d+)h)?(?:(\d+)m)?(?:(\d+(?:\.\d+)?)s)?$/.exec(text);
  if (!match) throw new Error(INVALID_TIME);
  const hours = Number(match[1] || 0);
  const minutes = Number(match[2] || 0);
  const seconds = Number(match[3] || 0);
  return hours * 3600 + minutes * 60 + seconds;
};

const validateSeconds = (seconds: number) => {
  if (!Number.isFinite(seconds)) throw new Error("Enter a finite start time.");
  return Math.round(seconds * 1000) / 1000;
};

/** Convert a supported timestamp; an empty field starts at zero. */
export const parseStartTime = (value: string): number => {
  const text = value.trim();
  if (!text) return 0;
  if (text.includes(":")) return validateSeconds(parseClockTime(text));
  if (/[hms]$/.test(text)) return validateSeconds(parseYoutubeTime(text));
  return validateSeconds(parseSeconds(text));
};

export const formatStartTime = (seconds: number): string => {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const [whole, fraction] = (seconds % 60).toFixed(3).split(".");
  const decimals = fraction.replace(/0+$/, "");
  const remainder = whole.padStart(2, "0") + (decimals ? `.${decimals}` : "");
  if (hours) return `${hours}:${String(minutes).padStart(2, "0")}:${remainder}`;
  return `${minutes}:${remainder}`;
};

export const youtubeStartTime = (value: string): string => {
  try {
    const url = new URL(value);
    const hosts = ["youtube.com", "www.youtube.com", "m.youtube.com", "youtu.be"];
    if (!hosts.includes(url.hostname)) return "";
    const fragment = new URLSearchParams(url.hash.slice(1));
    const timestamp =
      url.searchParams.get("t") ||
      url.searchParams.get("start") ||
      fragment.get("t");
    return timestamp ? formatStartTime(parseStartTime(timestamp)) : "";
  } catch {
    return "";
  }
};
